File: test/napi/lib/export.js
const assert = require("assert");

const addon = require("..");

describe("neon::export macro", () => {
  describe("globals", globals);
  describe("functions", functions);
});

function globals() {
  it("values", () => {
    assert.strictEqual(addon.NUMBER, 42);
    assert.strictEqual(addon.STRING, "Hello, World!");
    assert.strictEqual(addon.renamedString, "Hello, World!");
  });

  it("json", () => {
    assert.deepStrictEqual(addon.MESSAGES, ["hello", "neon"]);
    assert.deepStrictEqual(addon.renamedMessages, ["hello", "neon"]);
  });
}

function functions() {
  it("void function", () => {
    assert.strictEqual(addon.noArgsOrReturn(), undefined);
  });

  it("add - sync", () => {
    assert.strictEqual(addon.simpleAdd(1, 2), 3);
    assert.strictEqual(addon.renamedAdd(1, 2), 3);
  });

  it("add - task", async () => {
    const p1 = addon.addTask(1, 2);
    const p2 = addon.renamedAddTask(1, 2);

    assert.ok(p1 instanceof Promise);
    assert.ok(p2 instanceof Promise);

    assert.strictEqual(await p1, 3);
    assert.strictEqual(await p2, 3);
  });

  it("json sort", () => {
    const arr = ["b", "c", "a"];
    const expected = [...arr].sort();

    assert.deepStrictEqual(addon.jsonSort(arr), expected);
    assert.deepStrictEqual(addon.renamedJsonSort(arr), expected);
  });

  it("json sort - task", async () => {
    const arr = ["b", "c", "a"];
    const expected = [...arr].sort();
    const p1 = addon.jsonSortTask(arr);
    const p2 = addon.renamedJsonSortTask(arr);

    assert.ok(p1 instanceof Promise);
    assert.ok(p2 instanceof Promise);

    assert.deepStrictEqual(await p1, expected);
    assert.deepStrictEqual(await p2, expected);
  });

  it("can use context and handles", () => {
    const actual = addon.concatWithCxAndHandle("Hello,", " World!");
    const expected = "Hello, World!";

    assert.strictEqual(actual, expected);
  });

  it("error conversion", () => {
    const msg = "Oh, no!";
    const expected = new Error(msg);

    assert.throws(() => addon.failWithThrow(msg), expected);
  });

  it("tasks are concurrent", async () => {
    const time = 500;
    const sleep = (ms) => new Promise((r) => setTimeout(r, ms));
    const start = process.hrtime.bigint();

    await Promise.all([addon.sleepTask(time), sleep(time)]);

    const end = process.hrtime.bigint();
    const duration = end - start;

    // If `addon.sleepTask` blocks the thread, the tasks will run sequentially
    // and take a minimum of 2x `time`. Since they are run concurrently, we
    // expect the time to be closer to 1x `time`.
    const maxExpected = 2000000n * BigInt(time);

    assert.ok(duration < maxExpected);
  });

  it("can use generic Cx in exported functions", () => {
    assert.strictEqual(addon.numberWithCx(42), 42);
  });

  it("i32 parameters", () => {
    assert.strictEqual(addon.addI32(5, 3), 8);
    assert.strictEqual(addon.addI32(-10, 20), 10);
    assert.strictEqual(addon.addI32(-5, -3), -8);
    assert.strictEqual(addon.toI32(Infinity), 0);
    assert.strictEqual(addon.toI32(-Infinity), 0);
  });

  it("u32 parameters", () => {
    assert.strictEqual(addon.addU32(5, 3), 8);
    assert.strictEqual(addon.addU32(100, 200), 300);
    assert.strictEqual(addon.addU32(0, 42), 42);
    assert.strictEqual(addon.toU32(Infinity), 0);
    assert.strictEqual(addon.toU32(-Infinity), 0);
  });
}

File: test/napi/src/js/export.rs
use neon::{
    prelude::*,
    types::extract::{Boxed, Error},
};

#[neon::export]
const NUMBER: u8 = 42;

#[neon::export]
static STRING: &str = "Hello, World!";

#[neon::export(name = "renamedString")]
static RENAMED_STRING: &str = STRING;

#[neon::export(json)]
static MESSAGES: &[&str] = &["hello", "neon"];

#[neon::export(name = "renamedMessages", json)]
static RENAMED_MESSAGES: &[&str] = MESSAGES;

#[neon::export]
fn no_args_or_return() {}

#[neon::export]
fn simple_add(a: f64, b: f64) -> f64 {
    a + b
}

#[neon::export(name = "renamedAdd")]
fn rs_renamed_add(a: f64, b: f64) -> f64 {
    simple_add(a, b)
}

#[neon::export(task)]
fn add_task(a: f64, b: f64) -> f64 {
    simple_add(a, b)
}

#[neon::export(task, name = "renamedAddTask")]
fn rs_renamed_add_task(a: f64, b: f64) -> f64 {
    add_task(a, b)
}

#[neon::export(json)]
fn json_sort(mut items: Vec<String>) -> Vec<String> {
    items.sort();
    items
}

#[neon::export(json, name = "renamedJsonSort")]
fn rs_renamed_json_sort(items: Vec<String>) -> Vec<String> {
    json_sort(items)
}

#[neon::export(json, task)]
fn json_sort_task(items: Vec<String>) -> Vec<String> {
    json_sort(items)
}

#[neon::export(json, name = "renamedJsonSortTask", task)]
fn rs_renamed_json_sort_task(items: Vec<String>) -> Vec<String> {
    json_sort(items)
}

#[neon::export]
fn concat_with_cx_and_handle<'cx>(
    cx: &mut FunctionContext<'cx>,
    a: String,
    b: Handle<'cx, JsString>,
) -> Handle<'cx, JsString> {
    let b = b.value(cx);

    cx.string(a + &b)
}

#[neon::export]
fn fail_with_throw(msg: String) -> Result<(), Error> {
    fn always_fails(msg: String) -> Result<(), String> {
        Err(msg)
    }

    // `?` converts `String` into `Error`
    always_fails(msg)?;

    Ok(())
}

#[neon::export(task)]
fn sleep_task(ms: f64) {
    use std::{thread, time::Duration};

    thread::sleep(Duration::from_millis(ms as u64));
}

#[neon::export]
fn number_with_cx<'cx>(cx: &mut Cx<'cx>, n: f64) -> Handle<'cx, JsNumber> {
    cx.number(n)
}

#[neon::export]
fn simple_self(this: Handle<JsObject>) -> Handle<JsObject> {
    this
}

#[neon::export]
fn boxed_self(Boxed(this): Boxed<String>) -> String {
    this
}

#[neon::export]
fn boxed_string(s: String) -> Boxed<String> {
    Boxed(s)
}

#[neon::export]
fn add_i32(a: i32, b: i32) -> i32 {
    a + b
}

#[neon::export]
fn add_u32(a: u32, b: u32) -> u32 {
    a + b
}
